### PR TITLE
ARROW-5395: [C++] Utilize stream EOS in File format

### PR DESCRIPTION
**Pull request by jgm-ktg**
We currently do not write EOS at the end of a Message stream inside the File format.  As a result, the file cannot be parsed sequentially.  This change prepares for other implementations or future reference features that parse a File sequentially... i.e. without access to seek().

### CODE REVIEW

File: cpp/src/arrow/ipc/writer.cc [hunk 1]
```diff
@@ -1076,6 +1076,11 @@ class PayloadFileWriter : public internal::IpcPayloadWriter, protected StreamBoo
   }
 
   Status Close() override {
+    // Close stream for compatibility with sequential readers
+    RETURN_NOT_OK(UpdatePosition());
+    int32_t end_of_stream_marker = 0;
+    RETURN_NOT_OK(Write(&end_of_stream_marker, sizeof(int32_t)));
```
Comment (wesm) [inline cpp/src/arrow/ipc/writer.cc]: Can you move the `kEos` constexpr out of the `PayloadStreamWriter` to global scope and use that here? Then you don't need `end_of_stream_marker`

File: cpp/src/arrow/ipc/writer.cc [hunk 2]
```diff
@@ -1076,6 +1076,11 @@ class PayloadFileWriter : public internal::IpcPayloadWriter, protected StreamBoo
   }
 
   Status Close() override {
+    // Close stream for compatibility with sequential readers
+    RETURN_NOT_OK(UpdatePosition());
```
Comment (wesm) [inline cpp/src/arrow/ipc/writer.cc]: Don't think this postion update is needed

File: docs/source/format/IPC.rst [hunk 3]
```diff
@@ -104,7 +104,7 @@ Schematically we have: ::
 
     <magic number "ARROW1">
     <empty padding bytes [to 8 byte boundary]>
-    <STREAMING FORMAT>
+    <STREAMING FORMAT with EOS>
```
Comment (wesm) [inline docs/source/format/IPC.rst]: Someone should check if Java and JavaScript write the EOS

**Comment by wesm** [comment]
@jgm-ktg it's best to get out of the habit of using `git merge` and make pull requests from git branches. Not sure why CI did not trigger on your PR but it could be related to the merge somehow

File: docs/source/format/IPC.rst (hunk 3 above)
Comment (wesm) [reply docs/source/format/IPC.rst]: Had a look, it doesn't appear so https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileWriter.java#L67

Comment (jgm-ktg) [reply docs/source/format/IPC.rst]:
```         return this
+            ._writePadding(4); // eos bytes
             ._write(buffer) // Write the flatbuffer
             ._write(Int32Array.of(buffer.byteLength)) // then the footer size suffix
             ._writeMagic(); // then the magic suffix
```
and
```   protected void endInternal(WriteChannel out) throws IOException {
+    out.writeIntLittleEndian(0); // Write an EOS identifier for sequential readers
     long footerStart = out.getCurrentPosition();
```
? ...but I haven't figured out the testing yet

**Comment by wesm** [comment]
It occurs to me that we should probably add unit tests that the files can be iterated with a stream reader so this feature is not broken in the future

**Comment by jgm-ktg** [comment]
> It occurs to me that we should probably add unit tests that the files can be iterated with a stream reader so this feature is not broken in the future

```import pyarrow as pa
batch=pa.RecordBatch.from_arrays([ pa.array([1,None],type=pa.int32()) ], [ 'field1' ])

with open('/tmp/test.batch','wb') as sink:
    writer=pa.RecordBatchFileWriter(sink, batch.schema)
    writer.write_batch(batch)
    writer.close()

with open('/tmp/test.batch','rb') as source:
    reader=pa.ipc.open_stream(source.read()[8:])
    print(reader.read_pandas())
```

This appears to be a good test case?  If agreed I will try to add it to the Python testing, which will catch any C++ problem as well.

**Comment by wesm** [comment]
A C++ test case would be preferable but a Python test case is better than nothing

**Comment by jgm-ktg** [comment]
@wesm I backed the JS/Java changes out of this one and created separate JIRA entries.  I will add PRs to those as well once this one is merged... won't use my fork's master again for those, thanks for the tip.

**Review by wesm — APPROVED** [review]
+1. Thanks @jgm-ktg

**Pull request closed without merging**